radix-engine-common/src/data/manifest/custom_extension.rs

```diff
@@ -0,0 +1,58 @@
+use super::*;
+use crate::data::scrypto::*;
+use crate::*;
+use sbor::rust::prelude::*;
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum ManifestCustomExtension {}
+
+impl CustomExtension for ManifestCustomExtension {
+    const MAX_DEPTH: usize = MANIFEST_SBOR_V1_MAX_DEPTH;
+    const PAYLOAD_PREFIX: u8 = MANIFEST_SBOR_V1_PAYLOAD_PREFIX;
+
+    type CustomValueKind = ManifestCustomValueKind;
+    type CustomTraversal = ManifestCustomTraversal;
+    // NOTE: ManifestSbor is actually validated against Scrypto schemas
+    type CustomSchema = ScryptoCustomSchema;
+
+    fn custom_value_kind_matches_type_kind<L: SchemaTypeLink>(
+        custom_value_kind: Self::CustomValueKind,
+        type_kind: &TypeKind<<Self::CustomSchema as CustomSchema>::CustomTypeKind<L>, L>,
+    ) -> bool {
+        match custom_value_kind {
+            ManifestCustomValueKind::Address => matches!(
+                type_kind,
+                TypeKind::Custom(ScryptoCustomTypeKind::Reference)
+            ),
+            ManifestCustomValueKind::Bucket => {
+                matches!(type_kind, TypeKind::Custom(ScryptoCustomTypeKind::Own))
+            }
+            ManifestCustomValueKind::Proof => {
+                matches!(type_kind, TypeKind::Custom(ScryptoCustomTypeKind::Own))
+            }
+            // Vec<Own> buckets - will check more in the payload_validator
+            ManifestCustomValueKind::Expression => matches!(type_kind, TypeKind::Array { .. }),
```

Comment:
Can we at least verify the `element_type`, so anything built on top of `TypedTraverser` can have valid assumption about value-type-matching.

Similary for `Blob`.

Reply:
Yes - I agree with you on second thoughts.

It requires changing this method to take the schema reference - but I think that's the correct thing! :)